Opaque parameters

DougGregor opened the pull request:
Implement function parameters of the form `some P` be synthesizing an
implicit generic parameter whose requirements come from the opaque
type. We then map the opaque type back to the generic parameter, and
print as the opaque type. This allows us to write functions with
implicit generic parameters:

```swift
func f(_: some Collection) { }
```

which is equivalent to:

```swift
func f<T: Collection>(_: T) { }
```

where `T` is an otherwise-unused generic parameter name.

All of this is behind the experimental frontend flag
`-enable-experimental-opaque-parameters`.

Tracked via rdar://86032821

DougGregor committed:
Add frontend flag `-enable-experimental-opaque-parameters`

DougGregor committed:
Implement basic support for opaque parameters.

Implement function parameters of the form `some P` be synthesizing an
implicit generic parameter whose requirements come from the opaque
type. We then map the opaque type back to the generic parameter, and
print as the opaque type. This allows us to write functions with
implicit generic parameters:

```swift
func f(_: some Collection) { }
```

which is equivalent to:

```swift
func f<T: Collection>(_: some T) { }
```

where `T` is an otherwise-unused generic parameter name.

All of this is behind the experimental frontend flag
`-enable-experimental-opaque-parameters`.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

slavapestov commented:
Awesome!

> func f<T: Collection>(_: some T) { }

That `some` shouldn't be there right?

DougGregor commented:
> That `some` shouldn't be there right?

Right, thanks

DougGregor committed:
Augment GenericTypeParamDecl with bits indicating they came from opaque types

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor committed:
Suppress printing of implicit opaque generic parameters

When printing the generic parameters and requirements clauses of
declarations that involve opaque parameters, suppress the implicit
generic parameters created by the opaque parameters as well as the
generic requirements that involve said parameters.

DougGregor committed:
Make -enable-experimental-opaque-parameters a module interface option

That way, they can be used with Swift interfaces.

DougGregor committed:
Add a test mixing opaque parameters with primary associated types.

... beautiful.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor committed:
Allow opaque parameters in structural positions

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain
